Gameplay/entities/player/BulletController.cpp

```diff
@@ -1,100 +1,243 @@
+		// Check if its time to destroy
+		BulletStats& stats = _bullet_stats[i];
+		if (!stats.alive)
+		{
+			// Already dead bullet
+			continue;
+		}
+		if(stats.lifetime <= stats.elapsed_lifetime)
+		{
+			//	Disable if lifetime is ober
+			DeactivateBullet(i);
+			continue;
+		}
+		if (stats.current_charge < stats.charge_time)
+		{
+			// Do not launch yet
+			stats.current_charge += Time::DeltaTime();
+			bullet->GetTransform()->SetGlobalScale(float3(stats.GetChargedPercent()));
+			SetBulletTrajectory(i);
+			continue;
+		}
+
 		// Move bullet forward
-		current_position += delta_pos;
-		game_object->GetComponent<ComponentTransform>()->SetGlobalPosition(current_position);
+		UpdateBulletTrajectory(i);
 		// Check if it collides with an enemy
-		if (CheckCollisions())
+		if (CheckCollisions(i))
 		{
-			//	If it hits enemy bullets is destroyed
-			SceneManagement::Destroy(game_object);
-			return;
+			DeactivateBullet(i);
+			continue;
 		}
 
-		_lifetime -= Time::DeltaTime();
+		// If no collision just lower lifetime
+		stats.elapsed_lifetime += Time::DeltaTime();
+	}	
+}
+
+void Hachiko::Scripting::BulletController::UpdateBulletTrajectory(unsigned bullet_idx)
+{
+	GameObject* bullet = _bullets[bullet_idx];
+	BulletStats& stats = _bullet_stats[bullet_idx];
+	ComponentTransform* bullet_transform = bullet->GetTransform();
+	float3 current_position = bullet_transform->GetGlobalPosition();
+	stats.prev_position = current_position;
+	float3 new_position = current_position + stats.direction * stats.speed * Time::DeltaTime();
+	bullet_transform->SetGlobalPosition(new_position);
+}
+
+bool Hachiko::Scripting::BulletController::CheckCollisions(unsigned bullet_idx)
+{
+	// Process hit (if any) for closest enemy or obstacle
+
+	BulletStats& stats = _bullet_stats[bullet_idx];
+	GameObject* bullet = _bullets[bullet_idx];
+	LineSegment trajectory = LineSegment(stats.prev_position, bullet->GetTransform()->GetGlobalPosition());
+
+	float closest_enemy_hit = FLT_MAX;
+	EnemyController* hit_enemy = ProcessEnemies(bullet, stats.size, trajectory, closest_enemy_hit);
+	float closest_obstacle_hit = FLT_MAX;
+	CrystalExplosion* hit_obstacle = ProcessObstacles(bullet, stats.size, trajectory, closest_obstacle_hit);
+	
+	if (closest_enemy_hit < closest_obstacle_hit)
+	{
+		if (hit_enemy)
+		{
+			float3 knockback_dir = hit_enemy->GetGameObject()->GetTransform()->GetGlobalPosition() - bullet->GetTransform()->GetGlobalPosition();
+			hit_enemy->RegisterHit(stats.damage, knockback_dir);
+		}
+		return true;
 	}
+
+	if (hit_obstacle)
+	{
+		hit_obstacle->RegisterHit(stats.damage);
+		return true;
+	}
+
+	return false;
 }
 
-bool Hachiko::Scripting::BulletController::CheckCollisions()
+void Hachiko::Scripting::BulletController::ActivateBullet(unsigned bullet_idx)
 {
-	// Enemy collisions
-	GameObject* enemies = game_object->scene_owner->GetRoot()->GetFirstChildWithName("Enemies");
-	if (!enemies)	return false;
+	_bullet_stats[bullet_idx].alive = true;
```

Comment:
to manage bullets i identify then by their index in both bullet stats and bullets vector. I also send it to the player so it can request for its specific bullet to change state or something (and return -1 if no bullet could be generated)